src/Microsoft.Framework.DependencyInjection/Internal/ConcurrentDictionaryExtensions.cs

```diff
@@ -0,0 +1,43 @@
+﻿using System;
+
+namespace System.Collections.Concurrent
+{
+    internal static class ConcurrentDictionaryExtensions
+    {
+        // From https://github.com/dotnet/corefx/issues/394#issuecomment-69494764
+        // This lets us pass a state parameter allocation free GetOrAdd
+        internal static TValue GetOrAdd<TKey, TValue, TArg>(this ConcurrentDictionary<TKey, TValue> dictionary, TKey key, Func<TKey, TArg, TValue> valueFactory, TArg arg)
+        {
+            if (dictionary == null)
+            {
+                throw new ArgumentNullException(nameof(dictionary));
+            }
+
+            if (key == null)
+            {
+                throw new ArgumentNullException(nameof(key));
+            }
+
+            if (valueFactory == null)
+            {
+                throw new ArgumentNullException(nameof(valueFactory));
+            }
+
+            while (true)
+            {
+                TValue value;
+                if (dictionary.TryGetValue(key, out value))
+                {
+                    return value;
+                }
+
+                value = valueFactory(key, arg);
```

Comment:
You can make this code a little clearer by getting rid of the `while` and changing the `TryAdd` to `GetOrAdd(key, value)`. If your add fails, it's because someone was contending with you to also add a value and so you'll get the value they added.